### ISSUE TITLE
Changing default CloudWatch endpoint raises exception

### ISSUE THREAD

**Issue by yarons** [grafana/grafana#63420]
**What happened**:
I tried changing the CloudWatch endpoint in order to use a region specific one.

**What you expected to happen**:
It'll work and the query will work.

**How to reproduce it (as minimally and precisely as possible)**:
Instead of using the default, just type in a custom endpoint, either the global or region specific one.

**Anything else we need to know?**:

**Environment**:
- Grafana version: 9.3.6
- Data source type & version: 5.0.0
- OS Grafana is installed on: Docker on Amazon Linux 2
- User OS & Browser: Brave/Firefox
- Grafana plugins: Irrelevant
- Others: WebM Attached
[input_v0.webm](https://user-images.githubusercontent.com/406826/219948430-f2587ee1-1888-49b6-a8be-dfd7da6ece62.webm)

@LetsCreatProjects

**Comment by tonypowa** [comment]
:wave: Thanks for creating this issue, @yarons 

since you are getting a 404, it looks like it's a configuration problem from your side

before moving further, here's a [similar question](https://community.grafana.com/t/grafana-cloudwatch-plugin/76476)  posted in the community forum that might give you a hint where the problem is.

**Comment by yarons** [comment]
Hi @tonypowa and thank you for your quick answer :)
This is not the same case, I've added the `AWSXrayReadOnlyAccess` and there's no change.
If you can please look at the attached webm file you'll be able to see that when we're leaving the field blank everything work, typing in the internal endpoint doesn't and it should be completely seamless.

I couldn't find a way to look at the raw queries to see if there are any changes.
Thanks :)

**Comment by tonypowa** [comment]
thanks for the feedback and for sharing your research @yarons 

let's hear what the back-end squad think about it

**Comment by yarons** [comment]
Well according to AWS Support team, there are 2 different Endpoints:
monitoring.amazonaws.com
logs.amazonaws.com

Selecting either of them as the single endpoint will cause the other one to fail.
When leaving this field blank there's no such issue since the Grafana backend is using 2 different endpoints for this query which I can't specify as a user.

The screenshots from the support team:
### Logs
![image](https://user-images.githubusercontent.com/406826/222103017-ebec4be7-b64e-40bf-9a18-82fab49c7175.png)
### Monitoring
![image](https://user-images.githubusercontent.com/406826/222103049-7942a91d-ea90-44f9-af9e-cb426a575aed.png)

**Comment by idastambuk** [comment]
@yarons Hi, thanks for submitting the issue!
I see that you're getting the errors in the configuration editor and I've been able to reproduce them. 
While I do agree it is confusing to get an error for logs/metrics when testing the connection, this is somewhat expected, since the datasource checks both with the custom endpoint when `Save and test` is clicked. 
Since both of these datasources should be saved and possible to query, can you confirm that their respective queries are working fine when querying from a dashboard? 
In the meantime I will investigate if we can do something about making these errors and the UI more clear and seamless.

**Comment by yarons** [comment]
Hey @idastambuk,
Thank you for the quick response :)

I totally agree, it makes sense yet I"m also seeing error when trying to present the graph, otherwise it wouldn't be so critical.

Looking back at #31402 I can see that there was a proposal to allow specifying 2 distinct endpoints instead of 1.
I thought about tinkering with the code but I faced two issues:
1. I'm not sure how to replace the specific part of the code in my running instance.
2. Even after adding 2 endpoints, I'm not sure how to route the requests to the appropriate endpoint according to the respective URL.

I support separating the endpoints as I believe that it will avoid errors all at once (Leaving blank and specifying endpoint manually shouldn't lead to and error in any case unless a wrong URL was specified).

Thanks :)

**Comment by idastambuk** [comment]
Hi @yarons Are you able to provide more details on the errors in panels?
I've tried to reproduce the dashboard error by adding two Cloudwatch datasources and setting them up to hit the respective logs/metrics custom endpoint and I'm not getting an error. Can you try to describe the steps to reproduce? Thank you!

**Comment by yarons** [comment]
Sure @idastambuk.
I'm using a dashboard that is a derivative of the default EC2 dashboard, single internal endpoint configured (leaving the field blank doesn't cause these errors, Grafana version 9.3.6):
![grafana_errors](https://user-images.githubusercontent.com/406826/223095769-4f15a16a-50d8-4d6f-9e32-123e047b57a7.png)

**Comment by idastambuk** [comment]
Thank you @yarons for the info! Can you, by any chance, paste the request and response of the query endpoint from the network tab for this failed request? You should replace any sensitive data or IDs. 

Additionally, just to first confirm that errors aren't related to the datasource configuration - can you please confirm that you have [separated your Cloudwatch queries into two datasources](https://github.com/grafana/grafana/pull/31402#issuecomment-983883928) and are querying only metrics in this dashboard and that you've specified a monitoring endpoint for the datasource? 
If you are using the "Datasource" template variable in both metrics and logs panels, this won't work with the current implementation, as there should be two datasource variables, one for logs and one for metrics, like here for example: 
<img width="1161" alt="Screen Shot 2023-03-06 at 1 52 19 PM" src="https://user-images.githubusercontent.com/16140639/223115744-8668499d-9eb8-4cb3-95d2-cf33d8549ac2.png">

**Comment by yarons** [comment]
Hey @idastambuk,

I haven't since I'm not using the logs, only the metrics, it seems unrelated judging by the error.

This is the full error message:

```
unexpected error failed to call ec2:DescribeInstances, SerializationError: failed to unmarshal error message
	status code: 400, request id: 15d6cded-5aad-4451-999f-0685dc6b7f72
caused by: UnmarshalError: failed to unmarshal error message
	00000000  3c 45 72 72 6f 72 52 65  73 70 6f 6e 73 65 20 78  |<ErrorResponse x|
00000010  6d 6c 6e 73 3d 22 68 74  74 70 3a 2f 2f 77 65 62  |mlns="http://web|
00000020  73 65 72 76 69 63 65 73  2e 61 6d 61 7a 6f 6e 2e  |services.amazon.|
00000030  63 6f 6d 2f 41 57 53 46  61 75 6c 74 2f 32 30 30  |com/AWSFault/200|
00000040  35 2d 31 35 2d 30 39 22  3e 0a 20 20 3c 45 72 72  |5-15-09">.  <Err|
00000050  6f 72 3e 0a 20 20 20 20  3c 54 79 70 65 3e 53 65  |or>.    <Type>Se|
00000060  6e 64 65 72 3c 2f 54 79  70 65 3e 0a 20 20 20 20  |nder</Type>.    |
00000070  3c 43 6f 64 65 3e 49 6e  76 61 6c 69 64 41 63 74  |<Code>InvalidAct|
00000080  69 6f 6e 3c 2f 43 6f 64  65 3e 0a 20 20 20 20 3c  |ion</Code>.    <|
00000090  4d 65 73 73 61 67 65 3e  43 6f 75 6c 64 20 6e 6f  |Message>Could no|
000000a0  74 20 66 69 6e 64 20 6f  70 65 72 61 74 69 6f 6e  |t find operation|
000000b0  20 44 65 73 63 72 69 62  65 49 6e 73 74 61 6e 63  | DescribeInstanc|
000000c0  65 73 20 66 6f 72 20 76  65 72 73 69 6f 6e 20 32  |es for version 2|
000000d0  30 31 36 2d 31 31 2d 31  35 3c 2f 4d 65 73 73 61  |016-11-15</Messa|
000000e0  67 65 3e 0a 20 20 3c 2f  45 72 72 6f 72 3e 0a 20  |ge>.  </Error>. |
000000f0  20 3c 52 65 71 75 65 73  74 49 64 3e 31 35 64 36  | <RequestId>15d6|
00000100  63 64 65 64 2d 35 61 61  64 2d 34 34 35 31 2d 39  |cded-5aad-4451-9|
00000110  39 39 66 2d 30 36 38 35  64 63 36 62 37 66 37 32  |99f-0685dc6b7f72|
00000120  3c 2f 52 65 71 75 65 73  74 49 64 3e 0a 3c 2f 45  |</RequestId>.</E|
00000130  72 72 6f 72 52 65 73 70  6f 6e 73 65 3e 0a        |rrorResponse>.|

caused by: expected element type <Response> but have <ErrorResponse>null
```

HTML Only:
```html
<ErrorResponse xmlns="http://webservices.amazon.com/AWSFault/2005-15-09">
   .  
   <Error>
      .    
      <Type>Sender</Type>
      .    <Code>InvalidAction</Code>.    
      <Message>Could not find operation DescribeInstances for version 2016-11-15</Message>
      .  
   </Error>
   . 
   <RequestId>15d6cded-5aad-4451-999f-0685dc6b7f72</RequestId>
   .
</ErrorResponse>
.
```

According to AWS it's the wrong endpoint but I'm pretty certain that this is the right region so I'm not sure it's the same case.

**Comment by idastambuk** [comment]
@yarons You might be right that it is unrelated, the error seems to be coming from ec2. 
From a quick look at your screenshots, the region in the dashboard variable doesn't match the region in the configuration (us-east-1 vs. us-west-2), can you double check that both the variable and the Region input in datasource config have the same correct target region?

**Comment by idastambuk** [comment]
@yarons Thanks for checking this! It looks like it all checks out with regions. 
However, looking at the error message from aws, could it be that your datasource doesn't have sufficient permissions to access the data in AWS (Could not find operation DescribeInstances for version 2016-11-15)? Can you double check that your IAM has permissions listed [here](https://grafana.com/docs/grafana/latest/datasources/aws-cloudwatch/)? Thank you!

**Comment by yarons** [comment]
Hi @idastambuk my IAM role and the one you mentioned are identical, I've used the Metrics and Logs JSON to make sure that it won't fail in any case, it clearly doesn't work that way :)

I can even paste my role but I can assure you, it's identical.

**Comment by sarahzinger** [comment]
@yarons forgive me if I missed this in the back and forth, but I didn't see it. Were you able to follow the suggestion listed [here](https://github.com/grafana/grafana/pull/31402#issuecomment-983883928) as a temporary workaround to create 2 cloudwatch datasource instances in grafana, one for logs and one for metrics? Then you could use the logs datasource for logs queries and the metrics datasource for metrics queries?

Assuming that at least unblocks you from making the queries you need to, I think the next step would be for the Grafana team to investigate the possibility of putting 2 endpoint input fields, or explore if we can somehow reliably insert the word monitoring/logs in the right location in a custom endpoint.

**Comment by yarons** [comment]
Hi @sarahzinger and thank you for your help :)

I don't need the Logs endpoint, I only need the metrics.

My problem is that even though I'm only querying metrics I'm getting an error as mentioned [here](https://github.com/grafana/grafana/issues/63420#issuecomment-1455946760) so the only way for me to get a response is to leave the endpoint field empty.

Adding 2 optional endpoints is a good choice if I'm allowed to judge and will also mitigate the usage of the wrong endpoint (or the need to add two different data sources).

Thank you again :)

**Comment by yarons** [comment]
I'm pretty sure this is the default, we've used this one:
https://grafana.com/grafana/dashboards/617-aws-ec2/

All those parameters you mentioned are preconfigured.

**Comment by idastambuk** [comment]
Hi again @yarons!
I just imported the dashboard from this url and I'm getting some errors, even though they're not the same as the ones you provided. In order to debug this properly it would be great to have some more info that will help a lot with reproducing exactly what you have: 
1. Can you provide the url of the network request where you're getting [this response](https://github.com/grafana/grafana/issues/63420#issuecomment-1458363924) from? 
2. Are there panels that specifically stop working as soon as you switch the datasource from  one that has defined endpoint to the one that has this field blank? In case these are your custom panels - can you tell me if they use any of the template variables defined in this dashboard (instance name, instanceId, volumeId, and instanceType)? 
3. For each of these variables I have some values in the dropdowns - are your dropdowns empty even when you select something in Tag Name? If you have values - do the errors persist and are you seeing any values in the panels? 


Finally - I see the dashboard you linked hasn't been updated in a few years, so it's possible that there are some fields in the JSON that are not valid anymore. Are you able to make do with the sample EC2 dashboard imported from here?
<img width="700" alt="Screenshot 2023-03-22 at 4 50 15 PM" src="https://user-images.githubusercontent.com/16140639/226962108-3b91ec5c-1c37-43ce-9ba3-82cc6daa09dc.png">
 We're not seeing error there, so that one is probably more reliable in the meantime, while we debug https://grafana.com/grafana/dashboards/617-aws-ec2/
 
Thank you!

**Comment by yarons** [comment]
1. I've tried both Network Inspector in the browser and the docker logs, I can't find the actual request.
2. Only the datasource is working, all the others become blank.
3. Indeed, empty, reverting to blank endpoint fixes that.

Switched to the proposed dashboard and the errors are gone.
I think it's better to suggest that instead of the one I suggested earlier, thank you.

**Comment by MarcClusterman** [comment]
Facing a similar issue with respect to configuring this datasource to use the AWS FIPS endpoints (i.e. instead of `logs.us-east-1.amazonaws.com` use `logs-fips.us-east-1.amazonaws.com` and the same for `monitoring`. There appear to be 2 related issues:

1. This datasource doesn't appear to honor the environment variable `AWS_USE_FIPS_ENDPOINT`, which should cause `aws-sdk-go` to just do the right thing if no endpoint is specified for the session. My guess is this is because the endpoint is always explicitly set in https://github.com/grafana/grafana/blob/a1bc2272289ad453e196b78a434281eca925ca12/pkg/tsdb/cloudwatch/cloudwatch.go#L265
2. As mentioned above, this datasource uses 2 different AWS endpoints, so explicitly configuring an endpoint (FIPS or otherwise) causes either logs or metrics to fail

**Comment by sarahzinger** [comment]
@MarcClusterman Can you open a separate issue in this repo for number 1? Seems like a separate (but related!) issue that I'd love as much detail as possible in if you can! I quickly tried reproducing the issue and was not able to, but I am probably missing some additional repro steps.

For your second point I believe it should be resolved when we address https://github.com/grafana/grafana/issues/67850 although let me know if I'm missing anything. 

I'm going to close this issue for now as the original issue seems to have been mitigated with the other dashboard and the other work is being tracked elsewhere.
